### PR TITLE
Multithreading 6/N: Proxy calls to JavaScript functions from pthreads to main browser thread

### PR DESCRIPTION
**Pull request by juj**
This PR implements a mechanism where JavaScript library functions can be attributed with `foo__proxy: 'mode'`, where `mode` is one of `'main'`, `'async'`, `'main_gl'` or `'async_gl'`. This allows a pthread running in a web worker to either synchronously or asynchronously call out to JavaScript on the main browser thread.

The `_gl` variants decide proxying based on the currently active WebGL context for the calling thread. If the calling thread owns the GL context, then the call is not proxied. If the calling thread instead has activated a proxied context that is owned by the main browser thread, then the call is proxied. (This machinery is expanded on along with tests in a later PR)

This PR requires #5527 to land first in order to work.

### CODE REVIEW

File: src/library_pthread.js [hunk 1]
```diff
@@ -262,6 +262,42 @@ var LibraryPThread = {
         var worker = new Worker(pthreadMainJs);
 
         worker.onmessage = function(e) {
+          // TODO: Move the proxied call mechanism into a queue inside heap.
+          if (e.data.proxiedCall) {
+            var returnValue;
+            var funcTable = (e.data.func >= 0) ? proxiedFunctionTable : ASM_CONSTS;
+            var funcIdx = (e.data.func >= 0) ? e.data.func : (-1 - e.data.func);
+            PThread.currentProxiedOperationCallerThread = worker.pthread.threadInfoStruct; // Sometimes we need to backproxy events to the calling thread (e.g. HTML5 DOM events handlers such as emscripten_set_mousemove_callback()), so keep track in a globally accessible variable about the thread that initiated the proxying.
```
Comment (etienne02) [inline src/library_pthread.js]: Here I have an exception : `worker.pthread is undefined`, when a proxy call is lanch but the worker terminate and is clean.

File: src/jsifier.js [hunk 2]
```diff
@@ -100,6 +109,68 @@ function JSify(data, functionsOnly) {
...
+
+    function argsDict(num) { // ", p0: p0, p1: p1, p2: p2, p3: p3, p4: p4"
+      var s = '';
+      for(var i = 0; i < num; ++i) s += ', p' + i + ': p' + i;
+      return s;
+    }
+
+    // This is ad-hoc numbering scheme to map signatures to IDs, and must agree with call handler in src/library_pthread.js.
+    // Once proxied function calls no longer go through postMessage()s but instead in the heap, this will need to change, since int vs float will matter.
+    var functionCallOrdinal = sig.length + (sizeofReturn == 8 ? 20 : 0);
```
Comment (kripken) [inline src/jsifier.js]: what is magical about 8 and 20?

File: src/library_pthread.js (hunk 1 above)
Comment (juj) [reply src/library_pthread.js]: Any chance you'd be able to make a small test case?

**Comment by juj** [comment]
I'll need to write a full set of documentation about the updated pthreads support. I'd like to defer that to once these PRs land, so that I can write a coherent update of the current doc page at http://kripken.github.io/emscripten-site/docs/porting/pthreads.html

File: src/jsifier.js (hunk 2 above)
Comment (juj) [reply src/jsifier.js]: `sizeOfReturn` being 8 equals the case of a `double`, and the ID 20 is where functions with `double` return type start. The comment refers to the matching IDs at `src/library_pthread.js`: https://github.com/kripken/emscripten/pull/5531/files#diff-db41bea94577c2dd9b0eef0308b06cf9R272

Comment (juj) [reply src/jsifier.js]: Changed `sizeOfReturn == 8` test to `sig[0] == 'd'` to be more self-documenting.

File: src/library_pthread.js (hunk 1 above)
Comment (etienne02) [reply src/library_pthread.js]: I test to reproduce it in a small test case but for now, i'm not able. I will continue to try.

File: src/jsifier.js [hunk 3]
```diff
@@ -100,6 +109,68 @@ function JSify(data, functionsOnly) {
...
+
+    function argsDict(num) { // ", p0: p0, p1: p1, p2: p2, p3: p3, p4: p4"
+      var s = '';
+      for(var i = 0; i < num; ++i) s += ', p' + i + ': p' + i;
+      return s;
+    }
+
+    // This is ad-hoc numbering scheme to map signatures to IDs, and must agree with call handler in src/library_pthread.js.
+    // Once proxied function calls no longer go through postMessage()s but instead in the heap, this will need to change, since int vs float will matter.
+    var functionCallOrdinal = sig.length + (sig[0] == 'd' ? 20 : 0);
```
Comment (kripken) [inline src/jsifier.js]: i still think the 20 is bad - couldn't that become invalid if we change memory layout? or am i missing something?

at minimum lets put a big comment with XXX FIXME on this.

**Comment by kripken** [comment]
Bigger docs later sounds ok, but just for reviewing this PR, I could use some short explanations :)

File: src/jsifier.js (hunk 3 above)
Comment (juj) [reply src/jsifier.js]: The number `20` is more like a magic ID, it only needs to match in this spot in `src/jsifier.js` and then in `src/pthread-main.js` (in https://github.com/kripken/emscripten/pull/5531/files#diff-db41bea94577c2dd9b0eef0308b06cf9R272). There is nothing about the memory layout itself that would interact with this. Only limit to number 20 is if there's a function that needs proxying that would take in more than 20 input parameters, then we'd have to bump up the number.

Basically this is turning all these signature strings `'vii', 'viii', 'iii', 'dii'` into ID numbers, so that the receiving code knows what kind of function signature it should be calling.

**Comment by juj** [comment]
Edited the PR to remove the `main_gl` and `async_gl` proxying modes for now. This is because I'm looking at redoing the WebGL proxying via a more efficient method, by creating a WebGL thread. Depending on how that goes, I'll look at reintroducing `main_gl` and `async_gl`, whichever makes more sense.

**Comment by juj** [comment]
> Bigger docs later sounds ok, but just for reviewing this PR, I could use some short explanations :)

The `main` proxying mode queues the function call on the main browser thread, and the calling thread then synchronously waits for the function call to finish on the main thread, and gets the return value back.

The `async` proxying mode queues the function call on the main browser thread, but immediately continues to execute code without waiting for the function call to happen, or the return value, if the function had any.

**Comment by kripken** [comment]
I see, thanks. Why not call it sync and async, then, instead of main and async?

File: src/jsifier.js (hunk 3 above)
Comment (kripken) [reply src/jsifier.js]: I see, so it just cares about the # of params and whether the return value is float? maybe add a comment with that?

also, how about storing 20 in a const, and also asserting that the `sig.length` is shorter than it?

File: src/jsifier.js [hunk 4]
```diff
@@ -122,6 +193,11 @@ function JSify(data, functionsOnly) {
       // dependencies can be JS functions, which we just run
       if (typeof ident == 'function') return ident();
 
+      // don't process any special identifiers. These are looked up when processing the base name of the identifier.
+      if (ident.endsWith('__sig') || ident.endsWith('__proxy') || ident.endsWith('__asm') || ident.endsWith('__inline') || ident.endsWith('__deps') || ident.endsWith('__postset')) {
```
[the remark this answers is not in the document]
Comment (dschuff) [reply src/jsifier.js]: Actually I forgot that this is just using the same scheme we have, so ignore that comment :)

File: src/jsifier.js (hunk 3 above)
Comment (juj) [reply src/jsifier.js]: Good suggestions, improved this. Also while reading, found a yet unimplemented case in handling there, fixed that, and also changed the number from 20 to 32 to be more programmer-friendly :)

**Comment by juj** [comment]
> I see, thanks. Why not call it sync and async, then, instead of main and async?

Hmm, yeah, I first started with `main`, and then added `main_async`, and then thought that's too long, and shortened it to `async`. But `sync` and `async` sound like better, I'll change to that.

**Comment by juj** [comment]
Updated the PR, how does it look now?

File: src/jsifier.js [hunk 5]
```diff
@@ -100,6 +109,71 @@ function JSify(data, functionsOnly) {
...
+      for(var i = 0; i < num; ++i) s += ', p' + i;
+      return s;
+    }
+
+    var func = "function _emscripten_" + (sync ? '' : 'a') + 'sync_run_in_browser_thread_' + sig + '(func' + argsList(sig.length-1) + ') {\n';
+    if (sync) func += '  var waitAddress = Runtime.stackSave();\n';
+
+    function sizeofType(t) {
+      switch(t) {
+        case 'd':
```
Comment (kripken) [inline src/jsifier.js]: please do case bodies on same line, or `{,}`

Comment (juj) [reply src/jsifier.js]: check
